src/Protocol/Schema.php

```diff
@@ -0,0 +1,93 @@
+<?php
+declare(strict_types=1);
+
+namespace Lcobucci\Kafka\Protocol;
+
+use Lcobucci\Kafka\Protocol\Schema\Field;
+use function assert;
+use function is_array;
+
+/**
+ * Represents the fields present in a specific version of a protocol message (request or response)
+ */
+final class Schema extends Type
+{
+    /**
+     * @var Field[]
+     */
+    private $fields;
+
+    public function __construct(Field ...$fields)
+    {
+        $this->fields = $fields;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function write($data, Message $message): void
+    {
+        assert(is_array($data));
+
+        foreach ($this->fields as $field) {
+            $field->writeTo($data, $message);
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function read(Message $message): array
+    {
+        $structure = [];
+
+        foreach ($this->fields as $field) {
+            $structure[$field->name()] = $field->readFrom($message);
+        }
+
+        return $structure;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function sizeOf($data): int
+    {
+        assert(is_array($data));
+
+        $size = 0;
+
+        foreach ($this->fields as $field) {
+            $size += $field->sizeOf($data);
+        }
+
+        return $size;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function validate($data): void
+    {
+        $this->guardAgainstNull($data, 'array');
```

Comment:
I know they're not in this PR but why are these not `static`?

Reply:
Since this is already in the dynamic scope there's no gain in having it `static`...

Reply:
truth is, Luis doesn't like static stuff.

Reply:
I like, when it makes sense :+1: